pkg/trace/obfuscate/sql.go

```diff
@@ -206,7 +206,7 @@ func (o *Obfuscator) obfuscateSQL(span *pb.Span) {
 	if span.Resource == "" {
 		return
 	}
-	result, err := o.sql.obfuscate(span.Resource)
+	result, err := newSQLObfuscator().obfuscate(span.Resource)
```

Comment:
Why such a change in that PR?

Reply:
Because sharing its state between workers will create a race condition on the [tokenizer](https://github.com/DataDog/datadog-agent/blob/04c2ea1672ba8f5b6c659281be07fac368e2757f/pkg/trace/obfuscate/sql.go#L143).

Reply:
Won't it mean more allocations? Could we have one Obfuscator instance per worker?

Reply:
This is what I counted by looking at the code.

* 1 int
* 1 uint16
* 1 int64, 1 (empty) string, 1 int in the `strings.Reader`
* 4 pointers to `struct{}` (the filters)

And they don't escape the stack. I'd say it's negligible.

Do you think I've missed something?

Reply:
Right, it looks small (but still once per SQL span so a lot of calls) ; but let's be honest without complete benchmarks I can't tell if that's relevant or not.
 
So it is mostly a question of code organization ; if we go that way we should at least kill the dead code `Obfuscator.sql` (no longer used which is misleading).

Reply:
Or simpler: leave it to `sqlObfuscator.obfuscate` to instantiate the `tokenizer`?

Reply:
> Right, it looks small (but still once per SQL span so a lot of calls) ; but let's be honest without complete benchmarks I can't tell if that's relevant or not.

Correct. I shouldn't have even mentioned it without the numbers. Here is a comparison between the original version and this branch:

```
benchmark                             old ns/op     new ns/op     delta
BenchmarkTokenizer/Escaping/512-4     7733          7970          +3.06%
BenchmarkTokenizer/Grouping/199-4     4385          4375          -0.23%

benchmark                             old allocs     new allocs     delta
BenchmarkTokenizer/Escaping/512-4     87             91             +4.60%
BenchmarkTokenizer/Grouping/199-4     68             72             +5.88%

benchmark                             old bytes     new bytes     delta
BenchmarkTokenizer/Escaping/512-4     2992          3120          +4.28%
BenchmarkTokenizer/Grouping/199-4     1632          1760          +7.84%
```

> So it is mostly a question of code organization ; if we go that way we should at least kill the dead code Obfuscator.sql (no longer used which is misleading).

Done

> Or simpler: leave it to sqlObfuscator.obfuscate to instantiate the tokenizer?

Tried that, it's much smarter, more efficient, and easier to reason about. Also made all the fields local, since there is no point to have a shared state struct anymore. Now it's just a function. Results (compared to version in `master`) are:

```
benchmark                             old ns/op     new ns/op     delta
BenchmarkTokenizer/Escaping/512-4     7733          7940          +2.68%
BenchmarkTokenizer/Grouping/199-4     4385          4218          -3.81%

benchmark                             old allocs     new allocs     delta
BenchmarkTokenizer/Escaping/512-4     87             88             +1.15%
BenchmarkTokenizer/Grouping/199-4     68             69             +1.47%

benchmark                             old bytes     new bytes     delta
BenchmarkTokenizer/Escaping/512-4     2992          3008          +0.53%
BenchmarkTokenizer/Grouping/199-4     1632          1648          +0.98%
```

It's faster, and less memory is used now thanks to applying your comment. It's in the latest commit.

Reply:
Good catch @LotharSee